sklearn/utils/_array_api.py

```diff
@@ -74,22 +88,42 @@ def _check_array_api_dispatch(array_api_dispatch):
             )
 
 
-def device(x):
+def _single_array_device(array):
+    if isinstance(array, (numpy.ndarray, numpy.generic)) or not hasattr(
+        array, "device"
+    ):
+        return "cpu"
+    else:
+        return array.device
+
+
+def device(*array_list):
     """Hardware device the array data resides on.
 
     Parameters
     ----------
-    x : array
-        Array instance from NumPy or an array API compatible library.
+    *array_list : arrays
+        List of array instances from NumPy or an array API compatible library.
 
     Returns
     -------
     out : device
         `device` object (see the "Device Support" section of the array API spec).
     """
-    if isinstance(x, (numpy.ndarray, numpy.generic)):
-        return "cpu"
-    return x.device
+    if not array_list:
+        raise ValueError("At least one input array expected, got none.")
+
+    device_ = _single_array_device(array_list[0])
+
+    for array in array_list[1:]:
```

Comment:
```suggestion
    # Note: we only use equality tests in a loop over a list devices
    # instead of directly leveraging the unicity that would be induced
    # by a Python set datastructure: Python sets require hashable
    # members which is not guaranteed for Array API device objects. In
    # particular, this CuPy devices are not hashable at the time of
    # writing.
    for array in array_list[1:]:
```

Reply:
I'll add a test with mocks of non hashable devices.